src/util/positioning.ts

```diff
@@ -0,0 +1,100 @@
+export interface ElementPosition {
+  width: number;
+  height: number;
+  top: number;
+  left: number;
+}
+
+// previous version:
+// https://github.com/angular-ui/bootstrap/blob/07c31d0731f7cb068a1932b8e01d2312b796b4ec/src/position/position.js
+export class Positioning {
+  private getStyle(element: HTMLElement, prop: string): string { return window.getComputedStyle(element)[prop]; }
+
+  private isStaticPositioned(element: HTMLElement): boolean {
+    return (this.getStyle(element, 'position') || 'static') === 'static';
+  }
+
+  private parentOffsetElement(element: HTMLElement): HTMLElement {
+    let offsetParentEl: HTMLElement = element.parentElement || document.documentElement;
+    while (offsetParentEl && offsetParentEl !== document.documentElement) {
+      offsetParentEl = offsetParentEl.offsetParent as HTMLElement;
+    }
+    return offsetParentEl || document.documentElement;
+  }
+
+  position(element: HTMLElement): ElementPosition {
+    let elOffset: ElementPosition = this.offset(element);
+    let elBCR: ClientRect = element.getBoundingClientRect();
+    let parentOffset: ElementPosition = {width: 0, height: 0, top: 0, left: 0};
+    let offsetParentEl: HTMLElement = this.parentOffsetElement(element);
+
+    if (offsetParentEl !== document.documentElement) {
+      parentOffset = this.offset(offsetParentEl);
+      parentOffset.top += offsetParentEl.clientTop - offsetParentEl.scrollTop;
+      parentOffset.left += offsetParentEl.clientLeft - offsetParentEl.scrollLeft;
+    }
+
+    return {
+      width: elBCR.width || element.offsetWidth,
+      height: elBCR.height || element.offsetHeight,
+      top: elOffset.top - parentOffset.top,
+      left: elOffset.left - parentOffset.left
+    };
+  }
+
+  offset(element: HTMLElement): ElementPosition {
+    let elBcr: ClientRect = element.getBoundingClientRect();
+    return {
+      width: elBcr.width || element.offsetWidth,
+      height: elBcr.height || element.offsetHeight,
+      top: elBcr.top + (window.pageXOffset || document.documentElement.scrollTop),
+      left: elBcr.left + (window.pageXOffset || document.documentElement.scrollLeft)
+    };
+  }
+
+  positionElements(hostElement: HTMLElement, targetElement: HTMLElement, placement: string, appendToBody: boolean):
+      ElementPosition {
+    let hostElPosition: ElementPosition = appendToBody ? this.offset(hostElement) : this.position(hostElement);
+    let shiftWidth: any = {
+      left(): number{return hostElPosition.left;},
+      center(): number{return hostElPosition.left + hostElPosition.width / 2 - targetElement.offsetWidth / 2;},
+      right(): number{return hostElPosition.left + hostElPosition.width;}
```

Comment:
How come these have to be functions? Is this to readjust after the element has been rendered?